### PR TITLE
add LBFGSB class and unit test; modify all the interfaces of LBFGS,OWLQN

### PR DESCRIPTION
**Pull request by alice2008**
Change list:
1. add LBFGSB class and unit test
2. remove constrainMap in LBFGS class and update its unit test
3. remove constrainMap in OWLQN class and update its unit test

### CODE REVIEW

**Comment by alice2008** [comment]
@joshvfleming 
@ashelkovnykov

File: photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  protected def init(objectiveFunction: DiffFunction, initState: OptimizerState)(data: objectiveFunction.Data): Unit = {
+    var breezeDiffFunction: BreezeDiffFunction[DenseVector[Double]] = new BreezeDiffFunction[DenseVector[Double]] {
+      override def calculate(x: DenseVector[Double]): (Double, DenseVector[Double]) = {
+        var convertedCoefficients: objectiveFunction.Coefficients = objectiveFunction.convertFromVector(x)
+        var result = objectiveFunction.calculate(data, convertedCoefficients, normalizationContext)
+        (result._1, result._2.toDenseVector)
+      }
+    }
+    breezeOptimization = new BreezeOptimization (breezeDiffFunction, initState.coefficients)
+  }
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala]: Aside from the `breezeOptimizer` and the above function (`init`) everything looks mostly the same as in `LBFGS`. You should take a look at `OWLQN` extends `LBFGS` and do the same for `LBFGSB`.

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSTest.scala [hunk 2]
```diff
@@ -40,23 +40,18 @@ class LBFGSTest {
...
       // projection happens after it
-      Array(Some(Map[Int, (Double, Double)]()), Array(4.0)),
-      Array(Some(Map[Int, (Double, Double)](0 -> (3.0, 5.0))), Array(4.0)),
-      Array(Some(Map[Int, (Double, Double)](0 -> (Double.NegativeInfinity, 3.0))), Array(3.0)),
-      Array(Some(Map[Int, (Double, Double)](0 -> (5.0, Double.PositiveInfinity))), Array(5.0))
+      Array(Array(4.0))
     )
   }
 
   @Test(dataProvider = "dataProvider")
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSTest.scala]: Since there's only one test now, we may as well get rid of the data provider and just check that it optimizes towards `TestObjective.CENTROID`.

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import org.testng.annotations.{DataProvider, Test}
+
+class LBFGSBTest {
+
+  // TODO: set up boundary with non-zero expected loss value.
+  @DataProvider(name = "dataProvider")
+  def dataProvider(): Array[Array[Any]] = {
+    Array(
+      // Note that expected value here is the value of the function at the unconstrained optima since the
+      // projection happens after it
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: This comment does not make sense here, only in the old `OWLQN`.

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      Array(Array(Double.NegativeInfinity), Array(Double.PositiveInfinity), Array(4.0), 0.0),
+      Array(Array(-5.0), Array(5.0), Array(4.0), 0.0),
+      // set to -neg to 5.0 has convergence issue.
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: Can you tell me more about this?

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      Array(Array(Double.NegativeInfinity), Array(3.0), Array(3.0), 1.0),
+      Array(Array(-1.0), Array(Double.PositiveInfinity), Array(4.0), 0.0)
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: This case doesn't test anything new and should be replaced by the [5, inf) test case.

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    val normalizationContext = NoNormalization()
+    val normalizationContextBroadcast = mock(classOf[Broadcast[NormalizationContext]])
+    doReturn(normalizationContext).when(normalizationContextBroadcast).value
+
+    val lbfgsb = new LBFGSB(DenseVector(lowerBounds), DenseVector(upperBounds), normalizationContext = normalizationContextBroadcast)
+    val objective = new TestObjective
+    val trainingData = Array(LabeledPoint(0.0, CommonTestUtils.generateDenseVector(expectedCoef.length), 0.0, 0.0))
+    //if set initial coeffs to the mean of lower_bounds and upper_bounds, infinity values will go to NaN
+    val initialCoefficients = CommonTestUtils.generateDenseVector(expectedCoef.length)
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: Replace (+/-)inf with `Int.MaxValue`/`Int.MinValue`

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+      Array(Array(Double.NegativeInfinity), Array(3.0), Array(3.0), 1.0),
+      Array(Array(-1.0), Array(Double.PositiveInfinity), Array(4.0), 0.0)
+    )
+  }
+
+  @Test(dataProvider = "dataProvider")
+  def testLBFGSB(lowerBounds: Array[Double],
+                 upperBounds: Array[Double],
+                 expectedCoef: Array[Double],
+                 expectedValue: Double): Unit = {
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: See previous style comments.

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+package com.linkedin.photon.ml.optimization
+
+import breeze.linalg.DenseVector
+import com.linkedin.photon.ml.data.LabeledPoint
+import com.linkedin.photon.ml.normalization.{NoNormalization, NormalizationContext}
+import com.linkedin.photon.ml.test.{Assertions, CommonTestUtils}
+import org.apache.spark.broadcast.Broadcast
+import org.mockito.Mockito.{doReturn, mock}
+import org.testng.Assert.assertEquals
+import org.testng.annotations.{DataProvider, Test}
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: See previous style comments.

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+class LBFGSBTest {
+
+  // TODO: set up boundary with non-zero expected loss value.
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: I think this TODO is complete

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala (hunk 5 above)
Comment (alice2008) [reply photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: I want to pass this unit test but results in the optimizer error:
`Array(Array(5.0), Array(Double.PositiveInfinity), Array(5.0), 1.0)`
The reason is in breeze.LBFGSB, it will require the initial coeffs in the [lower_bound, upper_bound] range. However, the initial coeffs are generated with mean of 0 and std of 1 as in here:
`val initialCoefficients = CommonTestUtils.generateDenseVector(expectedCoef.length)`
It could be possible to generate the initialCoefficients with mean of (lower_bound+upper_bound)/2, however, it does not support Double.Infinity. 
`(Double.PositiveInfinity + 100.0)/2  //will be Infinity in scala`
One way is passing a really large value instead of Double.Infinity. For example:
`Array(Array(5.0), Array(1000.0), Array(5.0), 1.0)`

File: photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala (hunk 1 above)
Comment (alice2008) [reply photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala]: I was trying to just extend LBFGS class instead of extend Optimizer. There are several issues I am facing with:
1. `breezeDiffFunction` need to pass in `DenseVector[Double]` as type otherwise `breezeOptimizer` will report type mismatch when calling iterations function.
2. because of above 1, we need to redefine the nested class `BreezeOptimization` and `init` function. 
3. `var breezeOptimization` cannot be override because it is a var and we need to point it to the new nested class `BreezeOptimization`. 

I don't know what's the good way to extend LBFGS class here. We could talk more on this.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGS.scala [hunk 10]
```diff
@@ -67,9 +64,9 @@ class LBFGS(
     def next(state: OptimizerState): OptimizerState = {
       if (breezeStates.hasNext) {
         val breezeState = breezeStates.next()
-        // Project coefficients into constrained space, if any, before updating the state
+        // No-more constrain map
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGS.scala]: Don't forget to remove this comment before merge.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+  * @param normalizationContext The normalization context
+  * @param tolerance The tolerance threshold for improvement between iterations as a percentage of the initial loss
+  * @param maxNumIterations The cut-off for number of optimization iterations to perform.
+  * @param numCorrections The number of corrections used in the LBFGS update. Default 10. Values of numCorrections less
+  *                       than 3 are not recommended; large values of numCorrections will result in excessive computing
+  *                       time.
+  *                       Recommended:  3 < numCorrections < 10
+  *                       Restriction:  numCorrections > 0
+  * @param isTrackingState Whether to track intermediate states during optimization
+  */
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala]: Extra spaces in the comment blocks, should look like:
```
/**
 * Looks
 * like
 * this
 */
```

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.optimization
+
+import breeze.linalg.DenseVector
+import com.linkedin.photon.ml.data.LabeledPoint
+import com.linkedin.photon.ml.normalization.{NoNormalization, NormalizationContext}
+import com.linkedin.photon.ml.test.{Assertions, CommonTestUtils}
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: The breeze, spark, mockito, and testng imports should be in one import block together, and the photon imports should be in a second block below - take a look [here](https://github.com/databricks/scala-style-guide#imports).

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+class LBFGSBTest {
+
+  @DataProvider(name = "dataProvider")
+  def dataProvider(): Array[Array[Any]] = {
+    Array(
+      Array(Array(Double.NegativeInfinity), Array(Double.PositiveInfinity), Array(4.0), 0.0),
+      Array(Array(-5.0), Array(5.0), Array(4.0), 0.0),
+      Array(Array(Double.NegativeInfinity), Array(3.0), Array(3.0), 1.0),
+      Array(Array(-1.0), Array(Double.PositiveInfinity), Array(4.0), 0.0)
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: This test doesn't test anything new, it should be replaced by a test from [5, inf).

File: photon-api/src/main/scala/com/linkedin/photon/ml/optimization/OptimizerFactory.scala [hunk 14]
```diff
@@ -50,23 +50,21 @@ protected[ml] object OptimizerFactory {
           normalizationContext = normalizationContext,
           tolerance = config.tolerance,
           maxNumIterations = config.maximumIterations,
-          constraintMap = config.constraintMap,
           isTrackingState = isTrackingState)
 
-      case (OptimizerType.LBFGS, RegularizationType.L2 | RegularizationType.NONE) =>
+      case (OptimizerType.LBFGS, RegularizationType.L2 | RegularizationType.NONE) => {
```
Comment (ashelkovnykov) [inline photon-api/src/main/scala/com/linkedin/photon/ml/optimization/OptimizerFactory.scala]: The braces (`{`) aren't necessary here, let's get rid of them to remain consistent with the other case statements.

File: photon-client/src/main/scala/com/linkedin/photon/ml/Driver.scala [hunk 15]
```diff
@@ -49,6 +49,7 @@ import com.linkedin.photon.ml.supervised.model.{GeneralizedLinearModel, ModelTra
 import com.linkedin.photon.ml.supervised.regression.{LinearRegressionModel, PoissonRegressionModel}
 import com.linkedin.photon.ml.util.{IOUtils, PhotonLogger, Utils}
 
+
```
Comment (ashelkovnykov) [inline photon-client/src/main/scala/com/linkedin/photon/ml/Driver.scala]: Unnecessary whitespace

File: photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+          nextBreezeState.adjustedGradient,
+          state.iter+1
+        )
+      } else {
+        state
+      }
+    }
+  }
+
+  protected val BreezeOptimizer = new BreezeLBFGSB(lowerBounds, upperBounds, maxNumIterations, numCorrections, tolerance)
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala]: Should be `breezeOptimizer`. Missing the cast we discussed offline.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+  protected var breezeOptimization: BreezeOptimization = _
+
+  /**
+    * Initialize breeze optimization engine.
+    *
+    * @param objectiveFunction The objective function to be optimized
+    * @param initState The initial state of the optimizer, prior to starting optimization
+    * @param data The training data
+    */
+
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala]: See above comments

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+import org.mockito.Mockito.{doReturn, mock}
+import org.testng.Assert.assertEquals
+import org.testng.annotations.{DataProvider, Test}
+
+import com.linkedin.photon.ml.data.LabeledPoint
+import com.linkedin.photon.ml.normalization.{NoNormalization, NormalizationContext}
+import com.linkedin.photon.ml.test.{Assertions, CommonTestUtils}
+
+
+
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: Unnecessary whitespace.

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+class LBFGSBTest {
+
+  @DataProvider(name = "dataProvider")
+  def dataProvider(): Array[Array[Any]] = {
+    Array(
+      //Cannot use Double.PositiveInfinity or Double.NegativeInfinity because of mean used later
+      Array(Array(Int.MinValue*1.0), Array(Int.MaxValue*1.0), Array(4.0), 0.0),
+      Array(Array(-5.0), Array(5.0), Array(4.0), 0.0),
+      Array(Array(Int.MinValue*1.0), Array(3.0), Array(3.0), 1.0),
+      Array(Array(5.0), Array(Int.MaxValue*1.0), Array(5.0), 1.0)
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: Change `Int.MinValue` and `Int.MaxValue` to valid but smaller numbers (like `-10.0` and `10.0`) to avoid the non-convergence error.

File: photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+   * Under the hood, this adaptor uses an LBFGSB optimizer from Breeze to optimize functions.
+   * The DiffFunction is modified into a Breeze DiffFunction which the Breeze optimizer can understand.
+   */
+  //  protected class BreezeOptimization(
+  //      breezeDiffFunction: BreezeDiffFunction[DenseVector[Double]],
+  //      initCoefficients: Vector[Double])
+  //    extends BreezeOptimization(breezeDiffFunction.asInstanceOf[BreezeDiffFunction[Vector[Double]]], initCoefficients)
+  //
+  //  protected val BreezeOptimizer = new BreezeLBFGSB(lowerBounds, upperBounds, maxNumIterations, numCorrections, tolerance)
+  //
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala]: Some old code leftover here

File: photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+    asInstanceOf[FirstOrderMinimizer[Vector[Double], BreezeDiffFunction[Vector[Double]]]]
+
+  /**
+   * Initialize breeze optimization engine.
+   *
+   * @param objectiveFunction The objective function to be optimized
+   * @param initState         The initial state of the optimizer, prior to starting optimization
+   * @param data              The training data
+   */
+
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala]: Unnecessary whitespace

File: photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+    super.clearOptimizerInnerState()
+    breezeOptimization = _: BreezeBOptimization
+  }
+}
+
+object LBFGSB {
+  // TODO: Default tolerance and max # iterations should be in GameParams or GLMOptimizationConfiguration defaults
+  val DEFAULT_MAX_ITER = 100
+  val DEFAULT_NUM_CORRECTIONS = 10
+  val DEFAULT_TOLERANCE = 1.0E-7
```
Comment (ashelkovnykov) [inline photon-lib/src/main/scala/com/linkedin/photon/ml/optimization/LBFGSB.scala]: You can replace calls to these parameters with the ones from `LBFGS`. It'll be better that way since there will be an error if we ever replace them in `LBFGS` as the TODO comment suggests.

File: photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+class LBFGSBTest {
+
+  @DataProvider(name = "dataProvider")
+  def dataProvider(): Array[Array[Any]] = {
+    Array(
+      //Cannot use Double.PositiveInfinity or Double.NegativeInfinity because of mean used later
+      Array(Array(Int.MinValue*1.0), Array(Int.MaxValue*1.0), Array(4.0), 0.0),
+      Array(Array(-5.0), Array(5.0), Array(4.0), 0.0),
+      Array(Array(Int.MinValue.toDouble), Array(3.0), Array(3.0), 1.0),
+      Array(Array(5.0), Array(Int.MaxValue.toDouble), Array(5.0), 1.0)
```
Comment (ashelkovnykov) [inline photon-lib/src/test/scala/com/linkedin/photon/ml/optimization/LBFGSBTest.scala]: Change `Int.MinValue` and `Int.MaxValue` to valid but smaller numbers (like `-10.0` and `10.0`) to avoid the non-convergence error.

**Comment by ashelkovnykov** [comment]
Congratulations on your first PR!
